### DISCUSSION TITLE
Implementation of thermostat and WindowCovering in a bridgedDevice

### DISCUSSION THREAD

**Discussion by elausys** [matter-js/matter.js#996] [Q&A] [answered]
Hi,

I'm trying to create a bridgedDevice which includes different type of devices. It works well for some of the device types but I'm facing an issue to implement the thermostat and WindowCovering devices. 

See below the source code and resulting error. Could you help me to identify what could be the reason or provide an example of configuration for these types of devices ?

Thermstat device : 

![image](https://github.com/project-chip/matter.js/assets/152511534/b20e36ca-07d7-473e-8171-552e14a2c683)

Error : 

![image](https://github.com/project-chip/matter.js/assets/152511534/3b2580d5-27e7-4cd2-ba75-934b9f048fc4)

WindowCovering device : 

![image](https://github.com/project-chip/matter.js/assets/152511534/59fc7c19-60f6-450d-a5ba-d138d49fc6aa)

Error : 

![image](https://github.com/project-chip/matter.js/assets/152511534/8d7132b4-d951-4c59-99f2-d7b96008c207)

Thank you very much for your support !

**Comment by lauckhart** [comment, marked as answer]
Hi @elausys.  The error messages are cryptic here but the "conformance" expressions indicate the problem...

Matter conformance defines what you can do based on the features enabled on a cluster...  So for example, for the thermostat cluster you need to enable `Cooling` and `Heating` features to use `CoolingAndHeating` for `controlSequenceOfOperation`.  Conformance expression `[HEAT & COOL]` is saying `CoolingAndHeating` is allowed only if both cooling and heating are enabled.

For matter.js you can enable features using `with` on the behavior classes.  So e.g.:

```js
    ThermostatRequirements.ThermostatServer.with("Heating", "Cooling")
```

This enables the implementation in matter.js.  Then `CoolingAndHeating` is legal for `controlSequenceOfOperations`.

BTW if you want to avoid the constant "4" you can use matter.js's enum definition `Thermostat.ControlSequenceOfOperation.CoolingAndHeating`.

The window covering error is similar...  It's saying you can't have a shutter unless you enable `Lift` or `Tilt` features.

**Reply by Lovewife228** [reply]
it is correct answer!

**Reply by elausys** [reply]
Thank you very much for your fast response, it clarifies the issue indeed !

Best Regards,

**Comment by singh-io** [comment]
> Hi,
> 
> I'm trying to create a bridgedDevice which includes different type of devices. It works well for some of the device types but I'm facing an issue to implement the thermostat and WindowCovering devices.
> 
> See below the source code and resulting error. Could you help me to identify what could be the reason or provide an example of configuration for these types of devices ?
> 
> Thermstat device :
> 
> ![image](https://private-user-images.githubusercontent.com/152511534/346364278-b20e36ca-07d7-473e-8171-552e14a2c683.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA1NzIxNTksIm5iZiI6MTcyMDU3MTg1OSwicGF0aCI6Ii8xNTI1MTE1MzQvMzQ2MzY0Mjc4LWIyMGUzNmNhLTA3ZDctNDczZS04MTcxLTU1MmUxNGEyYzY4My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzEwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxMFQwMDM3MzlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04ZjU3N2U1NzAzMzFkMGU2OTFhYTE0MTkzNTkzOTQ5MmQ2YzlmNTcyY2ZhZmVhOTgxMjYxMjljYzAwNDdhYmM1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.JTLhQSNtjVoprsL6CklZkIyzK0ba3sZpynCLG9I7iUI)
> 
> Error :
> 
> ![image](https://private-user-images.githubusercontent.com/152511534/346364302-3b2580d5-27e7-4cd2-ba75-934b9f048fc4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA1NzIxNTksIm5iZiI6MTcyMDU3MTg1OSwicGF0aCI6Ii8xNTI1MTE1MzQvMzQ2MzY0MzAyLTNiMjU4MGQ1LTI3ZTctNGNkMi1iYTc1LTkzNGI5ZjA0OGZjNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzEwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxMFQwMDM3MzlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zN2ZlOWY0YmM4MDdiOTdiZTNjNGQ4MzIzM2NkOTdkY2E4YTRjYjIyOTRlYmJjODFkMDYzNTZjYTJiNmI4YTc3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.zM1NjD3zomMpm1sb3gh23M9ZqOo3lCH8YY9HWJ7CVNM)
> 
> WindowCovering device :
> 
> ![image](https://private-user-images.githubusercontent.com/152511534/346364334-59fc7c19-60f6-450d-a5ba-d138d49fc6aa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA1NzIxNTksIm5iZiI6MTcyMDU3MTg1OSwicGF0aCI6Ii8xNTI1MTE1MzQvMzQ2MzY0MzM0LTU5ZmM3YzE5LTYwZjYtNDUwZC1hNWJhLWQxMzhkNDlmYzZhYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzEwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxMFQwMDM3MzlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03MTc3MThiYmMyNzM2NDAwOWQ3ZjI2YmZiN2FmYzIzODg0YmY5ZWJhMTMxMmE0ZDYyMTU1MzAzYzU5NzNhNjlmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.7oX8DlwIXwIcG9nE80dcRMNG3y63rFwGUkEGCg4enk4)
> 
> Error :
> 
> ![image](https://private-user-images.githubusercontent.com/152511534/346364355-8d7132b4-d951-4c59-99f2-d7b96008c207.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA1NzIxNTksIm5iZiI6MTcyMDU3MTg1OSwicGF0aCI6Ii8xNTI1MTE1MzQvMzQ2MzY0MzU1LThkNzEzMmI0LWQ5NTEtNGM1OS05OWYyLWQ3Yjk2MDA4YzIwNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzEwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxMFQwMDM3MzlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01NjRjODAzZjk5YWU4ZjZhNjY5MDY5MDdjMGEzZDQyNDNiNDNiYjBhYzk1MmE1M2VhODU0NDZmMTM4MmUyOThiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.SSDG9o64GWwenw0jdE8sTfIXv3oC3mqQYOu8GuiyFTI)
> 
> Thank you very much for your support !

I am also trying something similar; trying to bridge lennox wifi thermostat  and airthings wave air quality sensor .. but i just started today with almost zero clue,  It would be a great help if you can share your code/example .. thanks. -g

**Reply by Apollon77** [reply]
@singh-io Please check the post, seems all images are broken ... and honestly  ... please post better "test" than images for code or error message because would be simply better readable

**Reply by singh-io** [reply]
I just replied to the original post sad 😢, was looking for some help. and surprisingly I can still see the images while writing this response, so it's not so easy to 'test', seems the GitHub markup is broken at multiple places.


-ta

**Reply by Lovewife228** [reply]
@singh-io Please share images with errors or code. I am also working with bridged devices which includes thermostat and window covering. I can help you.

**Reply by singh-io** [reply]
Can you point me to some examples for creating a bridged device?

**Reply by JimBuzbee** [reply]
There's a bridge example here: 

https://github.com/project-chip/matter.js/tree/main/packages/examples/src/device-bridge-onoff
